### PR TITLE
Fix tracker errors after a block reorganization

### PR DESCRIPTION
**Pull request by Nemanja0x**
The problem is when reorganization occurs on chain, so the parent block does not exist any more. function `handleReconcileImpl()` should check if the obtained block is `nil` (`parent == nil`):
`parent, err := t.provider.GetBlockByHash(block.ParentHash, false)`

There are some additional places in `blocktracker.go` and _tracker.go_ which should be handled also. They are introduces in this PR as well.

### CODE REVIEW

File: tracker/tracker.go [hunk 1]
```diff
@@ -314,10 +314,16 @@ func (t *Tracker) findAncestor(block, pivot *ethgo.Block) (uint64, error) {
 		block, err = t.provider.GetBlockByHash(block.ParentHash, false)
 		if err != nil {
 			return 0, err
+		} else if block == nil {
```
Comment (ferranbt) [inline tracker/tracker.go:317]: I do not think this is totally correct. What this code does is to check using the chain on hashes from the latest block you know. This scenario for example:
- A, B, C, D
Fork happens:
- A, B, C', D', E'

The loop will start in E' and go back trying to find the blocks until B since that is the common ancestor point. So it will do this block queries:
- Check parent of block E' => D'
- Check parent of block D' => C'
- Check parent of block C' => B (ancestor)

Now, the main question is, why would D' (for example) not be in the blockchain node? If the node has seen E', it must have D' and C'.

I think the reason might be because of the load balancer in cloud infrastructure, you might be hitting different nodes in the queries and getting different results.

What would happen with this change?

**Review by ferranbt — CHANGES_REQUESTED** [review]

File: tracker/tracker.go (hunk 1 above)
Comment (Nemanja0x) [reply tracker/tracker.go:317]: Let's say that block tracker has seen E' on a node and concurrently continues to read ancestors. While block tracker reads the next ancestor which is D', a reorganization can occur on the node. During the reorganization C', D', E' are removed. As as consequence block tracker will not be able to read D' because it does not exist anymore.

Comment (ferranbt) [reply tracker/tracker.go:317]: Mm.. but the blockchain does keep track of all the nodes and forks. So the other blocks should be available. Have you tested this branch already in master?

Comment (Nemanja0x) [reply tracker/tracker.go:317]: We have nil reference error in that place. The only thing that came to my mind is to handle it :)

image-20230419-133746

I'll run some more test on edge.

Comment (Nemanja0x) [reply tracker/tracker.go:317]: At that place there is already an error handling if the error is returned by the blockchain:

```
block, err = t.provider.GetBlockByHash(block.ParentHash, false)
if err != nil {
	return 0, err
} else if block == nil {
	// if block does not exist (for example reorg happened) GetBlockByNumber will return nil, nil
	return 0, fmt.Errorf("block with hash %s not found", block.ParentHash)
}
```
The idea is:  Blocktracker should behave the same if it didn't get the block. It does not matter if blockchain returned an error or the returned value is nil.

File: tracker/tracker.go [hunk 2]
```diff
@@ -882,3 +888,15 @@ func parseUint64orHex(str string) (uint64, error) {
 	}
 	return strconv.ParseUint(str, base, 64)
 }
+
+func getBlockByNumber(provider Provider, blockNumber uint64) (*ethgo.Block, error) {
```
Comment (ferranbt) [inline tracker/tracker.go:892]: Add this as a function of `Tracker`.

**Review by ferranbt — CHANGES_REQUESTED** [review]
Fix the comment and it should be good.

**Review by ferranbt — APPROVED** [review]
